File: etiquetas/generar_etiquetas.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Generador de etiquetas imprimibles para impresora termica Niimbot.

Lee los valores canonicos desde datos/circuitos.json (unica fuente de verdad)
y produce etiquetas/etiquetas-bjt.pdf:
  - Pagina EXACTA de 50 mm x 30 mm (141.732 x 85.039 pt).
  - Un circuito por pagina (3 paginas).
  - Solo negro puro sobre blanco. Reglas de 0.6-0.8 pt. Sin hairlines.
  - Ecuaciones de una sola linea "lhs = rhs = val" con subindices X_Y.
  - Dibujo manual con medicion de ancho (stringWidth) y auto-ajuste de
    tamano de fuente para garantizar que NADA se sale de los margenes.

Fuente: DejaVu Sans (TrueType, embebida) por su cobertura completa de
glifos tecnicos (beta, inf, ||, ~~, >=, ohm, micro, signo menos U+2212),
que la Helvetica Type1 estandar no renderiza de forma fiable.
"""

import json
import os
import re
import sys

from reportlab.lib.units import mm
from reportlab.pdfgen import canvas
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont
from reportlab.pdfbase.pdfmetrics import stringWidth

# ----------------------------------------------------------------------------
# Rutas
# ----------------------------------------------------------------------------
AQUI = os.path.dirname(os.path.abspath(__file__))
RAIZ = os.path.dirname(AQUI)
JSON_PATH = os.path.join(RAIZ, "datos", "circuitos.json")
PDF_PATH = os.path.join(AQUI, "etiquetas-bjt.pdf")

# ----------------------------------------------------------------------------
# Pagina (impresora termica Niimbot): 50 mm x 30 mm
# ----------------------------------------------------------------------------
PAGE_W = 50 / 25.4 * 72.0   # = 141.732 pt
PAGE_H = 30 / 25.4 * 72.0   # =  85.039 pt
PAGESIZE = (PAGE_W, PAGE_H)

MARGEN = 2.5                # margen ~2.5 pt (deja holgura para el bearing)
X0 = MARGEN
X1 = PAGE_W - MARGEN
USABLE_W = X1 - X0          # ~137.73 pt
TOP = PAGE_H - MARGEN

RULE_W = 0.7                # grosor de regla (>= 0.5 pt, sin hairlines)
SUB_FRAC = 0.72             # tamano del subindice relativo al cuerpo
SUB_DROP = 0.18             # cuanto baja el subindice (fraccion del tamano)
LEADING = 1.15              # interlineado

# ----------------------------------------------------------------------------
# Fuentes Unicode embebidas (DejaVu Sans regular + bold)
# ----------------------------------------------------------------------------
def _localizar_fuente(nombres):
    candidatos = []
    try:
        import matplotlib
        mpl = os.path.join(os.path.dirname(matplotlib.__file__),
                           "mpl-data", "fonts", "ttf")
        candidatos.append(mpl)
    except Exception:
        pass
    candidatos += [
        "/usr/share/fonts/dejavu",
        "/usr/share/fonts/truetype/dejavu",
        "/usr/share/fonts/TTF",
    ]
    for d in candidatos:
        for n in nombres:
            p = os.path.join(d, n)
            if os.path.isfile(p):
                return p
    return None

REG_PATH = _localizar_fuente(["DejaVuSans.ttf"])
BOLD_PATH = _localizar_fuente(["DejaVuSans-Bold.ttf"])
if not REG_PATH or not BOLD_PATH:
    sys.exit("ERROR: no se encontro DejaVuSans.ttf / DejaVuSans-Bold.ttf")

FONT = "DejaVuSans"
FONT_B = "DejaVuSans-Bold"
pdfmetrics.registerFont(TTFont(FONT, REG_PATH))
pdfmetrics.registerFont(TTFont(FONT_B, BOLD_PATH))

# ----------------------------------------------------------------------------
# Saneado / normalizacion de glifos
# ----------------------------------------------------------------------------
# DejaVu Sans soporta todos estos; aun asi normalizamos espacios raros.
def normaliza(s):
    if s is None:
        return ""
    s = s.replace(" ", " ")   # nbsp -> espacio normal
    s = s.replace(" ", " ")   # thin space
    return s

# ----------------------------------------------------------------------------
# Tokenizacion en "runs": (texto, es_subindice)
# Convierte "X_Y" en X con Y como subindice. El subindice abarca hasta el
# proximo separador (espacio, operador o parentesis).
# ----------------------------------------------------------------------------
SEP_SUB = set(" ()/+−-·*=,∥≈≥∞")  # donde termina un subindice

def tokeniza(texto):
    """Devuelve lista de runs [(texto, es_sub)] aplanando los subindices X_Y."""
    texto = normaliza(texto)
    runs = []
    buf = []          # texto base acumulado
    i = 0
    n = len(texto)
    while i < n:
        c = texto[i]
        if c == "_" and i + 1 < n:
            # cierra base pendiente
            if buf:
                runs.append(("".join(buf), False))
                buf = []
            # toma el subindice hasta el proximo separador
            j = i + 1
            # caso "X_(...)" -> subindice = contenido entre parentesis
            if texto[j] == "(":
                k = texto.find(")", j)
                if k != -1:
                    sub = texto[j + 1:k]
                    runs.append((sub, True))
                    i = k + 1
                    continue
            sub_chars = []
            while j < n and texto[j] not in SEP_SUB and texto[j] != "_":
                sub_chars.append(texto[j])
                j += 1
            if sub_chars:
                runs.append(("".join(sub_chars), True))
            i = j
        else:
            buf.append(c)
            i += 1
    if buf:
        runs.append(("".join(buf), False))
    return runs

def ancho_runs(runs, size):
    """Ancho total en pt de una lista de runs a un tamano de cuerpo dado."""
    w = 0.0
    for txt, es_sub in runs:
        fs = size * SUB_FRAC if es_sub else size
        w += stringWidth(txt, FONT, fs)
    return w

def dibuja_runs(c, x, y_base, runs, size):
    """Dibuja runs (con subindices) empezando en x; devuelve x final."""
    cx = x
    for txt, es_sub in runs:
        if es_sub:
            fs = size * SUB_FRAC
            c.setFont(FONT, fs)
            c.drawString(cx, y_base - size * SUB_DROP, txt)
            cx += stringWidth(txt, FONT, fs)
        else:
            c.setFont(FONT, size)
            c.drawString(cx, y_base, txt)
            cx += stringWidth(txt, FONT, size)
    return cx

# ----------------------------------------------------------------------------
# Construccion de las lineas de texto de cada etiqueta
# ----------------------------------------------------------------------------
def fmt_R(ohmios):
    """Resistencia en kohm con etiqueta compacta."""
    k = ohmios / 1000.0
    if abs(k - round(k)) < 1e-6:
        return f"{int(round(k))}kΩ"            # 4k ohm
    return f"{k:g}kΩ"

def fmt_C(faradios):
    return f"{faradios * 1e6:g}µF"             # micro F

def linea_datos(comp):
    """Construye la cadena 'Datos:' a partir de .componentes (kohm/uF/V/beta)."""
    partes = []
    # VCC primero
    if "VCC" in comp:
        partes.append(f"V_CC={comp['VCC']:g}V")
    # Resistencias en orden estable conocido
    for clave, etiq in [("R1", "R_1"), ("R2", "R_2"), ("RB", "R_B"),
                        ("RF", "R_F"), ("RC", "R_C"), ("RE", "R_E")]:
        if clave in comp:
            partes.append(f"{etiq}={fmt_R(comp[clave])}")
    if "beta" in comp:
        partes.append(f"β={comp['beta']:g}")
    # Condensadores
    for clave, etiq in [("C_in", "C_in"), ("C_out", "C_out"), ("C_E", "C_E")]:
        if clave in comp:
            partes.append(f"{etiq}={fmt_C(comp[clave])}")
    return " · ".join(partes)

def ecuacion(eq):
    """'lhs <rel> rhs = val' OMITIENDO subst. Detecta un operador relacional
    (≈, ≤, ≥, <, >) al inicio de rhs para no duplicar el '='."""
    lhs = normaliza(eq.get("lhs", ""))
    rhs_raw = eq.get("rhs", "").lstrip()
    val = normaliza(eq.get("val", ""))
    rel = "="
    for sym in ("≈", "≤", "≥", "<", ">"):
        if rhs_raw.startswith(sym):
            rel = sym
            rhs_raw = rhs_raw[len(sym):].strip()
            break
    rhs = normaliza(rhs_raw)
    return f"{lhs} {rel} {rhs} = {val}"

# Cada item es un dict que describe como dibujarlo.
# tipo: 'titulo' | 'regla' | 'etiqueta' | 'texto'
def construye_items(cir):
    items = []
    nombre = normaliza(cir["nombre"])
    subt = normaliza(cir.get("subtitulo", ""))
    items.append({"tipo": "titulo", "texto": nombre})
    if subt:
        items.append({"tipo": "subtitulo", "texto": subt})
    items.append({"tipo": "regla"})

    items.append({"tipo": "etiqueta", "texto": "Datos: " + linea_datos(cir["componentes"])})

    items.append({"tipo": "encabezado", "texto": "DC:"})
    for eq in cir.get("ecuaciones_dc", []):
        items.append({"tipo": "ecu", "texto": ecuacion(eq)})

    items.append({"tipo": "regla"})

    items.append({"tipo": "encabezado", "texto": "AC:"})
    for eq in cir.get("ecuaciones_ac", []):
        items.append({"tipo": "ecu", "texto": ecuacion(eq)})

    # recordatorio re = 26mV / IE
    items.append({"tipo": "recordatorio", "texto": "recordatorio: r_e = 26mV / I_E"})
    return items

# ----------------------------------------------------------------------------
# Medicion: dado un tamano base de cuerpo, calcula la altura total y si toda
# linea cabe en USABLE_W (envolviendo si hace falta). Devuelve (cabe, alto,
# lineas_render) donde lineas_render es lista de (item, runs).
# ----------------------------------------------------------------------------
def size_de(item, body):
    """Tamano de fuente segun el tipo de item, en funcion del cuerpo base."""
    t = item["tipo"]
    if t == "titulo":
        return body + 1.4, True   # negrita
    if t == "subtitulo":
        return body - 0.2, False
    if t == "encabezado":
        return body, True
    return body, False

def envuelve(runs, size, max_w):
    """Divide una lista de runs en varias lineas que quepan en max_w.
    El corte se hace en limites de palabra (espacios de runs base)."""
    # Primero re-expandimos a 'palabras' conservando subindices.
    palabras = []           # cada palabra = lista de runs
    actual = []
    for txt, es_sub in runs:
        if es_sub:
            actual.append((txt, True))
            continue
        # dividir el run base por espacios, conservando los espacios como
        # separadores de palabra
        trozos = re.split(r"(\s+)", txt)
        for tr in trozos:
            if tr == "":
                continue
            if tr.strip() == "":
                # espacio: cierra palabra
                if actual:
                    palabras.append(actual)
                    actual = []
            else:
                actual.append((tr, False))
    if actual:
        palabras.append(actual)

    # ahora empaqueta palabras en lineas
    lineas = []
    cur = []
    cur_w = 0.0
    space_w = stringWidth(" ", FONT, size)
    for pal in palabras:
        pw = ancho_runs(pal, size)
        extra = space_w if cur else 0.0
        if cur and cur_w + extra + pw > max_w:
            lineas.append(cur)
            cur = list(pal)
            cur_w = pw
        else:
            if cur:
                cur.append((" ", False))
                cur_w += space_w
            cur.extend(pal)
            cur_w += pw
    if cur:
        lineas.append(cur)
    if not lineas:
        lineas = [[("", False)]]
    return lineas

def maqueta(items, body):
    """Devuelve (cabe, alto_total, render) para un tamano de cuerpo base.
    render = lista de dicts: {tipo, kind('text'|'rule'), runs?, size?}"""
    render = []
    alto = 0.0
    max_overflow = 0.0
    for item in items:
        if item["tipo"] == "regla":
            # espacio antes + linea + espacio despues
            alto += body * 0.55
            render.append({"kind": "rule", "y_gap_before": 0.0})
            alto += body * 0.30
            continue
        size, bold = size_de(item, body)
        runs = tokeniza(item["texto"])
        # Importante: la negrita mide casi igual con DejaVu; medimos con FONT.
        lineas = envuelve(runs, size, USABLE_W)
        # registra overflow de la palabra mas ancha individual
        for ln in lineas:
            w = ancho_runs(ln, size)
            if w > USABLE_W:
                max_overflow = max(max_overflow, w - USABLE_W)
        for ln in lineas:
            render.append({"kind": "text", "tipo": item["tipo"],
                           "runs": ln, "size": size, "bold": bold})
            alto += size * LEADING
        # pequeno respiro tras titulo/encabezado
        if item["tipo"] in ("titulo",):
            alto += body * 0.10
    cabe = (alto <= (TOP - MARGEN)) and (max_overflow <= 0.01)
    return cabe, alto, render

# ----------------------------------------------------------------------------
# Dibujo de una pagina
# ----------------------------------------------------------------------------
def dibuja_pagina(c, items):
    # Busca el cuerpo mas grande que quepa (alto y ancho).
    body = 4.4
    render = None
    while body >= 2.6:
        cabe, alto, r = maqueta(items, body)
        if cabe:
            render = r
            break
        body -= 0.05
    if render is None:
        # ultimo recurso: usa el minimo y aun asi dibuja (no deberia ocurrir)
        _, _, render = maqueta(items, 2.6)
        body = 2.6

    c.setFillColorRGB(0, 0, 0)
    c.setStrokeColorRGB(0, 0, 0)

    y = TOP
    for el in render:
        if el["kind"] == "rule":
            y -= body * 0.55
            c.setLineWidth(RULE_W)
            c.setLineCap(0)
            c.line(X0, y, X1, y)
            y -= body * 0.30
            continue
        size = el["size"]
        y -= size  # bajamos al baseline de esta linea
        font = FONT_B if el["bold"] else FONT
        # dibujamos runs con la fuente correcta (bold/regular)
        cx = X0
        for txt, es_sub in el["runs"]:
            if es_sub:
                fs = size * SUB_FRAC
                c.setFont(font, fs)
                c.drawString(cx, y - size * SUB_DROP, txt)
                cx += stringWidth(txt, font, fs)
            else:
                c.setFont(font, size)
                c.drawString(cx, y, txt)
                cx += stringWidth(txt, font, size)
        y -= size * (LEADING - 1.0)
        if el["tipo"] == "titulo":
            y -= body * 0.10
    return body

# ----------------------------------------------------------------------------
# Main
# ----------------------------------------------------------------------------
def main():
    with open(JSON_PATH, "r", encoding="utf-8") as f:
        data = json.load(f)
    circuitos = data["circuitos"]

    c = canvas.Canvas(PDF_PATH, pagesize=PAGESIZE)
    c.setTitle("Etiquetas BJT - amplificadores")
    for cir in circuitos:
        items = construye_items(cir)
        body = dibuja_pagina(c, items)
        sys.stderr.write(f"  [{cir['id']}] {cir['nombre']}: cuerpo={body:.2f} pt\n")
        c.showPage()
    c.save()
    print(f"OK -> {PDF_PATH}  ({len(circuitos)} paginas, "
          f"{PAGE_W:.3f} x {PAGE_H:.3f} pt)")

if __name__ == "__main__":
    main()

File: etiquetas/generar_etiquetas.mjs
#!/usr/bin/env node
/*
 * Generador de etiquetas imprimibles Niimbot — versión Node (sin Python).
 *
 * Lee datos/circuitos.json (única fuente de verdad) y produce
 * etiquetas/etiquetas-bjt.pdf con pdfkit:
 *   - Página EXACTA de 50 mm × 30 mm (141.732 × 85.039 pt).
 *   - Un circuito por página (3 páginas).
 *   - Solo negro puro sobre blanco. Reglas de 0.7 pt (sin hairlines).
 *   - Ecuaciones de una línea "lhs = rhs = val" con subíndices X_Y.
 *   - Medición de ancho real (widthOfString) y auto-ajuste del tamaño de fuente
 *     para garantizar que NADA se sale de los márgenes.
 *
 * Fuente DejaVu Sans embebida (etiquetas/fonts/) por su cobertura de glifos
 * técnicos (β, ∞, ∥, ≈, Ω, µ, signo menos U+2212), igual que la versión previa.
 *
 * Uso:  node etiquetas/generar_etiquetas.mjs   (o: npm run etiquetas)
 */
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import PDFDocument from 'pdfkit'

const AQUI = path.dirname(fileURLToPath(import.meta.url))
const RAIZ = path.dirname(AQUI)
const JSON_PATH = path.join(RAIZ, 'datos', 'circuitos.json')
const PDF_PATH = path.join(AQUI, 'etiquetas-bjt.pdf')
const REG_PATH = path.join(AQUI, 'fonts', 'DejaVuSans.ttf')
const BOLD_PATH = path.join(AQUI, 'fonts', 'DejaVuSans-Bold.ttf')

// --- Página (impresora térmica Niimbot): 50 mm × 30 mm -----------------------
const PAGE_W = (50 / 25.4) * 72.0 // 141.732 pt
const PAGE_H = (30 / 25.4) * 72.0 //  85.039 pt
const MARGEN = 2.5
const X0 = MARGEN
const X1 = PAGE_W - MARGEN
const USABLE_W = X1 - X0
const USABLE_H = PAGE_H - 2 * MARGEN

const RULE_W = 0.7
const SUB_FRAC = 0.72 // tamaño del subíndice relativo al cuerpo
const SUB_DROP = 0.18 // cuánto baja el subíndice (fracción del cuerpo)
const LEADING = 1.15

const FREG = 'reg'
const FBOLD = 'bold'

// --- Normalización de glifos -------------------------------------------------
function normaliza(s) {
  if (s == null) return ''
  return String(s).replace(/ /g, ' ').replace(/ /g, ' ')
}

// Dónde termina un subíndice X_Y.
const SEP_SUB = new Set([...' ()/+−-·*=,∥≈≥∞'])

// Tokeniza "X_Y" en runs [texto, esSub], aplanando subíndices.
function tokeniza(texto) {
  texto = normaliza(texto)
  const runs = []
  let buf = ''
  let i = 0
  const n = texto.length
  while (i < n) {
    const c = texto[i]
    if (c === '_' && i + 1 < n) {
      if (buf) { runs.push([buf, false]); buf = '' }
      let j = i + 1
      if (texto[j] === '(') {
        const k = texto.indexOf(')', j)
        if (k !== -1) {
          runs.push([texto.slice(j + 1, k), true])
          i = k + 1
          continue
        }
      }
      let sub = ''
      while (j < n && !SEP_SUB.has(texto[j]) && texto[j] !== '_') { sub += texto[j]; j++ }
      if (sub) runs.push([sub, true])
      i = j
    } else {
      buf += c
      i++
    }
  }
  if (buf) runs.push([buf, false])
  return runs
}

// --- Medición de ancho (mide siempre con la regular, como la versión previa) --
function anchoRuns(doc, runs, size) {
  let w = 0
  doc.font(FREG)
  for (const [txt, esSub] of runs) {
    const fs = esSub ? size * SUB_FRAC : size
    doc.fontSize(fs)
    w += doc.widthOfString(txt)
  }
  return w
}

// --- Formato de componentes --------------------------------------------------
function g(x) {
  // Equivalente a "{x:g}": quita ceros sobrantes.
  return String(parseFloat(Number(x).toPrecision(6)))
}
function fmtR(ohm) {
  const k = ohm / 1000
  return (Math.abs(k - Math.round(k)) < 1e-6 ? String(Math.round(k)) : g(k)) + 'kΩ'
}
function fmtC(far) { return g(far * 1e6) + 'µF' }

function lineaDatos(comp) {
  const partes = []
  if ('VCC' in comp) partes.push(`V_CC=${g(comp.VCC)}V`)
  for (const [clave, etiq] of [['R1', 'R_1'], ['R2', 'R_2'], ['RB', 'R_B'], ['RF', 'R_F'], ['RC', 'R_C'], ['RE', 'R_E']]) {
    if (clave in comp) partes.push(`${etiq}=${fmtR(comp[clave])}`)
  }
  if ('beta' in comp) partes.push(`β=${g(comp.beta)}`)
  for (const [clave, etiq] of [['C_in', 'C_in'], ['C_out', 'C_out'], ['C_E', 'C_E']]) {
    if (clave in comp) partes.push(`${etiq}=${fmtC(comp[clave])}`)
  }
  return partes.join(' · ')
}

// "lhs <rel> rhs = val" omitiendo subst; detecta operador relacional en rhs.
function ecuacion(eq) {
  const lhs = normaliza(eq.lhs || '')
  let rhsRaw = (eq.rhs || '').replace(/^\s+/, '')
  const val = normaliza(eq.val || '')
  let rel = '='
  for (const sym of ['≈', '≤', '≥', '<', '>']) {
    if (rhsRaw.startsWith(sym)) { rel = sym; rhsRaw = rhsRaw.slice(sym.length).trim(); break }
  }
  return `${lhs} ${rel} ${normaliza(rhsRaw)} = ${val}`
}

// --- Items de cada etiqueta --------------------------------------------------
function construyeItems(cir) {
  const items = []
  items.push({ tipo: 'titulo', texto: normaliza(cir.nombre) })
  const subt = normaliza(cir.subtitulo || '')
  if (subt) items.push({ tipo: 'subtitulo', texto: subt })
  items.push({ tipo: 'regla' })
  items.push({ tipo: 'etiqueta', texto: 'Datos: ' + lineaDatos(cir.componentes) })
  items.push({ tipo: 'encabezado', texto: 'DC:' })
  for (const eq of cir.ecuaciones_dc || []) items.push({ tipo: 'ecu', texto: ecuacion(eq) })
  items.push({ tipo: 'regla' })
  items.push({ tipo: 'encabezado', texto: 'AC:' })
  for (const eq of cir.ecuaciones_ac || []) items.push({ tipo: 'ecu', texto: ecuacion(eq) })
  items.push({ tipo: 'recordatorio', texto: 'recordatorio: r_e = 26mV / I_E' })
  return items
}

function sizeDe(item, body) {
  switch (item.tipo) {
    case 'titulo': return [body + 1.4, true]
    case 'subtitulo': return [body - 0.2, false]
    case 'encabezado': return [body, true]
    default: return [body, false]
  }
}

// Divide runs en líneas que quepan en maxW (corte en límites de palabra).
function envuelve(doc, runs, size, maxW) {
  const palabras = []
  let actual = []
  for (const [txt, esSub] of runs) {
    if (esSub) { actual.push([txt, true]); continue }
    for (const tr of txt.split(/(\s+)/)) {
      if (tr === '') continue
      if (tr.trim() === '') { if (actual.length) { palabras.push(actual); actual = [] } }
      else actual.push([tr, false])
    }
  }
  if (actual.length) palabras.push(actual)

  doc.font(FREG).fontSize(size)
  const spaceW = doc.widthOfString(' ')
  const lineas = []
  let cur = []
  let curW = 0
  for (const pal of palabras) {
    const pw = anchoRuns(doc, pal, size)
    const extra = cur.length ? spaceW : 0
    if (cur.length && curW + extra + pw > maxW) {
      lineas.push(cur); cur = [...pal]; curW = pw
    } else {
      if (cur.length) { cur.push([' ', false]); curW += spaceW }
      cur.push(...pal); curW += pw
    }
  }
  if (cur.length) lineas.push(cur)
  return lineas.length ? lineas : [[['', false]]]
}

// Maqueta a un cuerpo dado: devuelve {cabe, alto, render}.
function maqueta(doc, items, body) {
  const render = []
  let alto = 0
  let maxOverflow = 0
  for (const item of items) {
    if (item.tipo === 'regla') {
      alto += body * 0.55
      render.push({ kind: 'rule' })
      alto += body * 0.30
      continue
    }
    const [size, bold] = sizeDe(item, body)
    const runs = tokeniza(item.texto)
    const lineas = envuelve(doc, runs, size, USABLE_W)
    for (const ln of lineas) {
      const w = anchoRuns(doc, ln, size)
      if (w > USABLE_W) maxOverflow = Math.max(maxOverflow, w - USABLE_W)
    }
    for (const ln of lineas) {
      render.push({ kind: 'text', tipo: item.tipo, runs: ln, size, bold })
      alto += size * LEADING
    }
    if (item.tipo === 'titulo') alto += body * 0.10
  }
  const cabe = alto <= USABLE_H && maxOverflow <= 0.01
  return { cabe, alto, render }
}

// Dibuja una etiqueta cuyo borde superior está en originY (coords pdfkit, y↓).
function dibujaEtiqueta(doc, items, originY) {
  let body = 4.4
  let render = null
  while (body >= 2.6) {
    const m = maqueta(doc, items, body)
    if (m.cabe) { render = m.render; break }
    body -= 0.05
  }
  if (!render) { render = maqueta(doc, items, 2.6).render; body = 2.6 }

  doc.fillColor('black').strokeColor('black')
  let y = originY + MARGEN // cursor de baseline (y crece hacia abajo)
  for (const el of render) {
    if (el.kind === 'rule') {
      y += body * 0.55
      doc.lineWidth(RULE_W).lineCap('butt').moveTo(X0, y).lineTo(X1, y).stroke()
      y += body * 0.30
      continue
    }
    const size = el.size
    y += size // baja al baseline de esta línea
    const font = el.bold ? FBOLD : FREG
    let cx = X0
    for (const [txt, esSub] of el.runs) {
      if (esSub) {
        const fsz = size * SUB_FRAC
        doc.font(font).fontSize(fsz).text(txt, cx, y + size * SUB_DROP, { baseline: 'alphabetic', lineBreak: false })
        cx += doc.widthOfString(txt)
      } else {
        doc.font(font).fontSize(size).text(txt, cx, y, { baseline: 'alphabetic', lineBreak: false })
        cx += doc.widthOfString(txt)
      }
    }
    y += size * (LEADING - 1.0)
    if (el.tipo === 'titulo') y += body * 0.10
  }
  return body
}

const HOJA_PATH = path.join(AQUI, 'etiquetas-hoja.pdf') // hoja de contacto (3 apiladas)
const GAP = 8 // separación entre etiquetas en la hoja de contacto (pt)

function escribe(doc, ruta) {
  return new Promise((resolve) => {
    const s = fs.createWriteStream(ruta)
    doc.pipe(s)
    doc.end()
    s.on('finish', resolve)
  })
}

// --- Main --------------------------------------------------------------------
async function main() {
  const data = JSON.parse(fs.readFileSync(JSON_PATH, 'utf8'))
  const circuitos = data.circuitos

  // (1) PDF de etiquetas: una por página (50×30 mm).
  const doc = new PDFDocument({ size: [PAGE_W, PAGE_H], margin: 0, autoFirstPage: false })
  doc.registerFont(FREG, REG_PATH)
  doc.registerFont(FBOLD, BOLD_PATH)
  doc.info.Title = 'Etiquetas BJT - amplificadores'
  for (const cir of circuitos) {
    doc.addPage({ size: [PAGE_W, PAGE_H], margin: 0 })
    const body = dibujaEtiqueta(doc, construyeItems(cir), 0)
    process.stderr.write(`  [${cir.id}] ${cir.nombre}: cuerpo=${body.toFixed(2)} pt\n`)
  }
  await escribe(doc, PDF_PATH)
  console.log(`OK -> ${PDF_PATH}  (${circuitos.length} páginas, ${PAGE_W.toFixed(3)} x ${PAGE_H.toFixed(3)} pt)`)

  // (2) Hoja de contacto: las 3 etiquetas apiladas en una página, con separadores
  //     negros sobre blanco. Sirve para previsualizar todas a la vez (preview-todas).
  const hojaH = PAGE_H * circuitos.length + GAP * (circuitos.length - 1)
  const hoja = new PDFDocument({ size: [PAGE_W, hojaH], margin: 0, autoFirstPage: true })
  hoja.registerFont(FREG, REG_PATH)
  hoja.registerFont(FBOLD, BOLD_PATH)
  circuitos.forEach((cir, k) => {
    const originY = k * (PAGE_H + GAP)
    dibujaEtiqueta(hoja, construyeItems(cir), originY)
    if (k < circuitos.length - 1) {
      const yy = originY + PAGE_H + GAP / 2
      hoja.lineWidth(0.7).strokeColor('black').moveTo(0, yy).lineTo(PAGE_W, yy).stroke()
    }
  })
  await escribe(hoja, HOJA_PATH)
  console.log(`OK -> ${HOJA_PATH}  (hoja de contacto, ${PAGE_W.toFixed(3)} x ${hojaH.toFixed(3)} pt)`)
}

main()
